### PR TITLE
[Feat]: IPO Feature

### PR DESCRIPTION
**Pull request by Hedhav**
The IPO Feature is designed to introduce newer stocks after the game has already started. All IPO companies will be added to the IpoStocks table. The stocks will be issued in bundles, known as slots.

Timeline of the IPO Feature:

-  The OpenIpoBidding RPC is called from the admin panel.
-  Users will be able to place bids on these IPO stocks (currently the limit is only 1 slot per user) and they will be be able to cancel placed bids. 
- Users will only be able to place and cancel bids when the market is open. The bid amount is moved from the user's cash to reservedCash.
 <br>

- The CloseIpoBidding RPC is called from the admin panel.
- Users will no longer be able to bid on IPO stocks (isBiddable is set to false).
- The IpoStock is added to the stock table.
- The Slots (StocksPerSlot number of stocks per bid) will be allotted depending on over-subscription or under-subscription condition:
- - Over-subscription (total no of bids > available slots)
- - - (Available slots) number of bids are randomly selected to get the IPO, rest get a refund.
- -  Under-subscription (total no of bids <= available slots)
- - - All bids are fulfilled, remaining slots are "deleted", i.e., never enter the market.
- - The prices (ListingPrice) and StocksInMarket are updated accordingly for the new stock.
- - Users who are allotted a slot will get their reservedCash deducted and get the stocks added to their portfolio
- - Users whose bids are unfulfilled get a refund (the bid amount is moved back from reservedCash to cash)

### CODE REVIEW

File: grpcapi/actionservice/OrderActionService.go [hunk 1]
```diff
@@ -149,6 +149,108 @@ func (d *dalalActionService) PlaceOrder(ctx context.Context, req *actions_pb.Pla
...
+	}
+
+	if models.IsUserBlocked(userId) {
+		return makeError(actions_pb.PlaceIpoBidResponse_UserBlockedError, "Your account has been blocked due to malpractice.")
+	}
+
+	var IpoBidId uint32
+	var err error
+
+	IpoBidId, err = models.CreateIpoBid(userId, req.StockId, req.SlotQuantity, req.SlotPrice)
```
Comment (siva2204) [inline grpcapi/actionservice/OrderActionService.go:184]: ### 
```suggestion

	IpoBidId, err := models.CreateIpoBid(userId, req.StockId, req.SlotQuantity, req.SlotPrice)
```

**Review by siva2204 — CHANGES_REQUESTED** [review]
todo : integrate with data streams (can be done at last)

File: models/Ipo.go [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+		ListingPrice = IpoStock.StockPrice
+	} else if subscriptionRatio <= 1.20 && subscriptionRatio >= 1.05 {
+		ListingPrice = IpoStock.StockPrice * 105 / 100
+	} else if subscriptionRatio <= 0.95 && subscriptionRatio >= 0.80 {
+		ListingPrice = IpoStock.StockPrice * 95 / 100
+	} else if subscriptionRatio >= 1.20 {
+		ListingPrice = IpoStock.StockPrice * 110 / 100
+	} else { // if subscriptionRatio <= 0.80
+		ListingPrice = IpoStock.StockPrice * 90 / 100
+	}
```
[the remark this answers is not in the document]
Comment (Hedhav) [reply models/Ipo.go:288]: but wouldn't these allotment transactions affect the final price of the stock if i save listing price first? the plan was to decide the listing price which will be its value when the market opens after allotment

File: models/Ipo.go [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+				l.Error(err)
+				return err
+			}
+
+			if err := db.Save(ipoBid).Error; err != nil {
+				l.Error(err)
+				return err
+			}
+
+			IpoStocksInMarket = uint64(len(openIpoBids))
```
[the remark this answers is not in the document]
Comment (Hedhav) [reply models/Ipo.go:212]: IpoStocksInMarket = total ipo stocks if oversubscribed, and total stocks bidded for if undersubscribed (rest are deleted)

correction, this line should be:   IpoStocksInMarket = uint64(uint32(len(openIpoBids)) * IpoStock.StocksPerSlot)

File: models/Ipo.go [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+		return err
+	}
+
+	cost := int64(uint64(IpoStock.SlotQuantity) * IpoStock.SlotPrice)
+	// NewStockId = newStock.Id // Will this be defined or do i have to check db to get it?
+
+	if subscriptionRatio <= 1.00 {
+		for _, ipoBid := range openIpoBids {
+			// allot 1 slot worth of stocks to userid
+			AllotIpoTransaction := GetTransactionRef(ipoBid.UserId, newStock.Id, IpoAllotmentTransaction, 0, int64(IpoStock.SlotQuantity*IpoStock.StocksPerSlot), 0, -cost, 0)
```
[the remark this answers is not in the document]
Comment (Hedhav) [reply models/Ipo.go:194]: his reserved cash is being depleted, so isnt it correct? (cost is positive)

File: models/Ipo.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (siva2204) [reply models/Ipo.go:288]: transaction needs stock price, we're doing transaction before calculating the price, which doesnt add up anything to user protfolio

**Review by siva2204 — DISMISSED** [review]

**Review by satyamurti — APPROVED** [review]

**Pull request merged**